Liberty 18.0.0.1 startup issues with Arabic locale

When starting my Liberty server and running using locale ar_DZ.UTF-8, the Liberty server fails to start with these messages in the log file:

Launching domainconsole (WebSphere Application Server 18.0.0.1/wlp-1.0.20.cl180120180309-2209) on IBM J9 VM, version 8.0.5.16 - pxa6480sr5fp16-20180524_01(SR5 FP16) (ar_DZ)

CWWKE0005E: The runtime environment could not be launched.
CWWKE0018E: An exception occurred while launching the runtime environment: java.lang.NumberFormatException: For input string: ""
com.ibm.ws.kernel.boot.LaunchException: Caught unexpected exception java.lang.NumberFormatException: For input string: ""
    at com.ibm.ws.kernel.boot.internal.KernelBootstrap.rethrowException(KernelBootstrap.java:530)
    at com.ibm.ws.kernel.boot.internal.KernelBootstrap.go(KernelBootstrap.java:218)
    at com.ibm.ws.kernel.boot.Launcher.handleActions(Launcher.java:246)
    at com.ibm.ws.kernel.boot.Launcher.createPlatform(Launcher.java:121)
    at com.ibm.ws.kernel.boot.cmdline.EnvCheck.main(EnvCheck.java:59)
    at com.ibm.ws.kernel.boot.cmdline.EnvCheck.main(EnvCheck.java:35)
Caused by: java.lang.NumberFormatException: For input string: ""
    at java.lang.NumberFormatException.forInputString(NumberFormatException.java:76)
    at java.lang.Long.parseLong(Long.java:613)
    at java.lang.Long.parseLong(Long.java:643)
    at com.ibm.ws.logging.collector.BurstDateFormat.format(BurstDateFormat.java:105)
    at com.ibm.ws.logging.collector.DateFormatHelper.formatTime(DateFormatHelper.java:65)
    at com.ibm.ws.logging.internal.impl.BaseTraceFormatter.messageLogFormat(BaseTraceFormatter.java:395)
    at com.ibm.ws.logging.internal.impl.JsonTraceService.publishLogRecord(JsonTraceService.java:337)
    at com.ibm.ws.logging.internal.impl.BaseTraceService.info(BaseTraceService.java:307)
    at com.ibm.websphere.ras.Tr.info(Tr.java:632)
    at com.ibm.ws.logging.internal.impl.IncidentImpl.log(IncidentImpl.java:264)
    at com.ibm.ws.logging.internal.impl.BaseFFDCService.log(BaseFFDCService.java:239)
    at com.ibm.ws.logging.internal.impl.BaseFFDCService.processException(BaseFFDCService.java:216)
    at com.ibm.ws.ffdc.FFDCFilter.processException(FFDCFilter.java:94)
    at com.ibm.ws.kernel.launch.internal.LauncherDelegateImpl.launchFramework(LauncherDelegateImpl.java:103)
    at com.ibm.ws.kernel.boot.internal.KernelBootstrap.go(KernelBootstrap.java:212)
    ... 4 more

--- a/dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/BurstDateFormat.java
+++ b/dev/com.ibm.ws.logging/src/com/ibm/ws/logging/collector/BurstDateFormat.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2017 IBM Corporation and others.
+ * Copyright (c) 2017, 2018 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -13,6 +13,7 @@ package com.ibm.ws.logging.collector;
 import java.text.DateFormat.Field;
 import java.text.FieldPosition;
 import java.text.SimpleDateFormat;
+import java.util.Date;
 
 /**
  * Formats dates that are relatively close to one another.
@@ -66,12 +67,18 @@ public class BurstDateFormat {
     private boolean reset = false;
 
     /**
+     * Tracks whether the format is not valid. If true, the underlying SimpleDateFormat is used
+     */
+    protected boolean invalidFormat = false;
+
+    /**
      * Constructs a BurstDateFormat
-     * 
+     *
      * @param formatter
      */
     public BurstDateFormat(SimpleDateFormat formatter) {
         this.formatter = formatter;
+        setup();
     }
 
     /**
@@ -83,6 +90,41 @@ public class BurstDateFormat {
         if (!formatter.toPattern().equals(pattern)) {
             formatter.applyPattern(pattern);
             reset = true;
+            invalidFormat = false;
+            setup();
+        }
+    }
+
+    /**
+     * Setup the date formatter, determine if the format is valid
+     */
+    protected void setup() {
+        try {
+            StringBuffer refTime = new StringBuffer();
+
+            // Get the positions of the millisecond digits
+            String pattern = formatter.toLocalizedPattern();
+            if (pattern.lastIndexOf('S') - pattern.indexOf('S') != 2) {
+                invalidFormat = true;
+            }
+            formatter.format(new Date().getTime(), refTime, position);
+
+            // Should be redundant check with above but check again
+            // just in case some locale expands the milliseconds field
+            if (position.getEndIndex() - position.getBeginIndex() != 3) {
+                invalidFormat = true;
+            }
+            String str = refTime.substring(position.getBeginIndex(), position.getEndIndex());
+
+            // Make sure we are using ascii digits (The loop could be unwrapped)
+            for (char a : str.toCharArray()) {
+                if (a < 48 || a > 57) {
+                    invalidFormat = true;
+                    break;
+                }
+            }
+        } catch (Exception e) {
+            invalidFormat = true;
         }
     }
 
@@ -93,33 +135,51 @@ public class BurstDateFormat {
      * @return
      */
     public String format(long timestamp) {
-        long delta = timestamp - refTimestamp;
 
-        // If we need to reformat
-        if (delta >= pdiff || delta < ndiff || reset) {
-            reset = false;
-            StringBuffer refTime = new StringBuffer();
-            refTimestamp = timestamp;
-            formatter.format(timestamp, refTime, position);
-
-            refMilli = Long.parseLong(refTime.substring(position.getBeginIndex(), position.getEndIndex()));
-            refBeginning = refTime.substring(0, position.getBeginIndex());
-            refEnding = refTime.substring(position.getEndIndex());
-
-            pdiff = 1000 - refMilli;
-            ndiff = -refMilli;
-            return refTime.toString();
-        } else {
-            StringBuffer sb = new StringBuffer();
-            long newMilli = delta + refMilli;
-            if (newMilli >= 100)
-                sb.append(refBeginning).append(newMilli).append(refEnding);
-            else if (newMilli >= 10)
-                sb.append(refBeginning).append('0').append(newMilli).append(refEnding);
-            else
-                sb.append(refBeginning).append("00").append(newMilli).append(refEnding);
-            return sb.toString();
-
+        // If the format is unknown, use the default formatter.
+        if (invalidFormat) {
+            return formatter.format(timestamp);
         }
+
+        try {
+            long delta = timestamp - refTimestamp;
+
+            // If we need to reformat
+            if (delta >= pdiff || delta < ndiff || reset) {
+                reset = false;
+                StringBuffer refTime = new StringBuffer();
+                refTimestamp = timestamp;
+                formatter.format(timestamp, refTime, position);
+
+                refMilli = Long.parseLong(refTime.substring(position.getBeginIndex(), position.getEndIndex()));
+
+                refBeginning = refTime.substring(0, position.getBeginIndex());
+                refEnding = refTime.substring(position.getEndIndex());
+
+                pdiff = 1000 - refMilli;
+                ndiff = -refMilli;
+                return refTime.toString();
+            } else {
+                StringBuffer sb = new StringBuffer();
+                long newMilli = delta + refMilli;
+                if (newMilli >= 100)
+                    sb.append(refBeginning).append(newMilli).append(refEnding);
+                else if (newMilli >= 10)
+                    sb.append(refBeginning).append('0').append(newMilli).append(refEnding);
+                else
+                    sb.append(refBeginning).append("00").append(newMilli).append(refEnding);
+                return sb.toString();
+
+            }
+        } catch (Exception e) {
+            // Throw FFDC in case anything goes wrong
+            // Still generate the date via the SimpleDateFormat
+            invalidFormat = true;
+            return formatter.format(timestamp);
+        }
+    }
+
+    public SimpleDateFormat getSimpleDateFormat() {
+        return formatter;
     }
 }

--- a/dev/com.ibm.ws.logging_test/test/com/ibm/ws/logging/collector/BurstDateFormatterTest.java
+++ b/dev/com.ibm.ws.logging_test/test/com/ibm/ws/logging/collector/BurstDateFormatterTest.java
@@ -1,0 +1,195 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.logging.collector;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.GregorianCalendar;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Random;
+
+import org.junit.Test;
+
+public class BurstDateFormatterTest {
+
+    private static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+
+    @Test
+    public void checkAllLocalesTest() {
+        for (Locale locale : Locale.getAvailableLocales()) {
+            SimpleDateFormat simple = (SimpleDateFormat) getFormatter(DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, locale));
+            BurstDateFormat burst = new BurstDateFormat(simple);
+            compareFormat(burst, simple);
+        }
+    }
+
+    @Test
+    public void checkAllLocalesWithAllOffsetTest() {
+        GregorianCalendar cal = new GregorianCalendar();
+        cal.set(Calendar.MILLISECOND, 143);
+        Date initialDate = cal.getTime();
+
+        HashMap<Locale, BurstDateFormat> localeMap = new HashMap<Locale, BurstDateFormat>();
+
+        // Initialize the BurstDateFormatter
+        for (Locale locale : Locale.getAvailableLocales()) {
+            SimpleDateFormat simple = (SimpleDateFormat) getFormatter(DateFormat.getDateTimeInstance(DateFormat.SHORT, DateFormat.MEDIUM, locale));
+            BurstDateFormat burst = new BurstDateFormat(simple);
+            localeMap.put(locale, burst);
+            compareFormat(locale, initialDate, burst, simple);
+        }
+
+        // Loop through every digit to make sure it is correct
+        for (int i = 0; i < 1000; i++) {
+            cal.set(Calendar.MILLISECOND, i);
+            Date checkDate = cal.getTime();
+            for (Locale locale : localeMap.keySet()) {
+                compareFormat(locale, checkDate, localeMap.get(locale), localeMap.get(locale).getSimpleDateFormat());
+            }
+        }
+
+    }
+
+    @Test
+    public void reverseTimeTest() {
+        SimpleDateFormat simple = new SimpleDateFormat(DATE_FORMAT);
+        BurstDateFormat burst = new BurstDateFormat(simple);
+        GregorianCalendar cal = new GregorianCalendar();
+
+        cal.set(Calendar.MILLISECOND, 892);
+        Date checkDate = cal.getTime();
+        compareFormat(checkDate, burst, simple);
+
+        cal.set(Calendar.MILLISECOND, 384);
+        checkDate = cal.getTime();
+        compareFormat(checkDate, burst, simple);
+
+        cal.set(Calendar.MILLISECOND, 192);
+        checkDate = cal.getTime();
+        compareFormat(checkDate, burst, simple);
+
+        cal.add(Calendar.MILLISECOND, -500);
+        checkDate = cal.getTime();
+        compareFormat(checkDate, burst, simple);
+        assertTrue("BurstDateFormat failed to identify valid format", !burst.invalidFormat);
+    }
+
+    @Test
+    public void noMilliSecondsTest() {
+        SimpleDateFormat simple = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
+        BurstDateFormat burst = new BurstDateFormat(simple);
+        compareFormat(burst, simple);
+    }
+
+    @Test
+    public void regularDateFormatTest() {
+        SimpleDateFormat simple = new SimpleDateFormat(DATE_FORMAT);
+        BurstDateFormat burst = new BurstDateFormat(simple);
+        compareFormat(burst, simple);
+        assertTrue("BurstDateFormat failed to identify valid format", !burst.invalidFormat);
+    }
+
+    @Test
+    public void randomDatesTest() {
+        SimpleDateFormat simple = new SimpleDateFormat(DATE_FORMAT);
+        BurstDateFormat burst = new BurstDateFormat(simple);
+
+        Random rand = new Random();
+        for (long i = 0; i < 5000; i++) {
+            compareFormat(new Date(rand.nextLong()), burst, simple);
+            assertTrue("BurstDateFormat failed to identify valid format", !burst.invalidFormat);
+        }
+    }
+
+    @Test
+    public void multipleMillisecondFormatTest() {
+        SimpleDateFormat simple = new SimpleDateFormat("SSS ss SSS ss");
+        BurstDateFormat burst = new BurstDateFormat(simple);
+        GregorianCalendar cal = new GregorianCalendar();
+        Date date = cal.getTime();
+        compareFormat(date, burst, simple);
+        assertTrue("BurstDateFormat failed to identify invalid format", burst.invalidFormat);
+
+        // Test offset
+        cal.add(Calendar.MILLISECOND, 74);
+        compareFormat(date, burst, simple);
+    }
+
+    /**
+     * Compares the BurstDateFormat with SimpleDateFormat
+     */
+    private void compareFormat(BurstDateFormat burst, SimpleDateFormat simple) {
+        Date date = new Date();
+        compareFormat(date, burst, simple);
+    }
+
+    /**
+     * Compares the BurstDateFormat with SimpleDateFormat
+     */
+    private void compareFormat(Date date, BurstDateFormat burst, SimpleDateFormat simple) {
+        compareFormat("BurstDateFormat failed to match SimpleDateFormat", date, burst, simple);
+    }
+
+    /**
+     * Compares the BurstDateFormat with SimpleDateFormat
+     */
+    private void compareFormat(Locale locale, Date date, BurstDateFormat burst, SimpleDateFormat simple) {
+        compareFormat("BurstDateFormat failed to match SimpleDateFormatfor locale " + locale, date, burst, simple);
+    }
+
+    /**
+     * Compares the BurstDateFormat with SimpleDateFormat
+     */
+    private void compareFormat(String msg, Date date, BurstDateFormat burst, SimpleDateFormat simple) {
+        long timestamp = (date).getTime();
+        String burstString = burst.format(timestamp);
+        String simpleString = simple.format(date);
+        assertEquals(msg, burstString, simpleString);
+    }
+
+    /**
+     * A partial copy of DateFormatHelper.getDateFormat so that we can work with different locales
+     */
+    public DateFormat getFormatter(DateFormat formatter) {
+        String pattern;
+        int patternLength;
+        int endOfSecsIndex;
+        if (formatter instanceof SimpleDateFormat) {
+            // Retrieve the pattern from the formatter, since we will need to
+            // modify it.
+            SimpleDateFormat sdFormatter = (SimpleDateFormat) formatter;
+            pattern = sdFormatter.toPattern();
+            // Append milliseconds and timezone after seconds
+            patternLength = pattern.length();
+            endOfSecsIndex = pattern.lastIndexOf('s') + 1;
+            String newPattern = pattern.substring(0, endOfSecsIndex) + ":SSS z";
+            if (endOfSecsIndex < patternLength)
+                newPattern += pattern.substring(endOfSecsIndex, patternLength);
+            // 0-23 hour clock (get rid of any other clock formats and am/pm)
+            newPattern = newPattern.replace('h', 'H');
+            newPattern = newPattern.replace('K', 'H');
+            newPattern = newPattern.replace('k', 'H');
+            newPattern = newPattern.replace('a', ' ');
+            newPattern = newPattern.trim();
+            sdFormatter.applyPattern(newPattern);
+            formatter = sdFormatter;
+        } else {
+            formatter = new SimpleDateFormat("dd/MMM/yyyy HH:mm:ss:SSS z");
+        }
+        return formatter;
+    }
+}

